stubs/CoreGenericFunctions.phpstub

```diff
@@ -725,6 +725,26 @@ function join($separator, array $array = []): string
 /**
  * @psalm-pure
  *
+ * @param non-empty-string $separator
+ *
+ * @return (
+ *     $limit is int<min, -2>
+ *         ? array<empty, empty>
+ *         : (
+ *             $limit is int<-1, 1>
+ *                 ? array{string}
+ *                 : (
+ *                     $limit is 2
+ *                         ? array{0: string, 1?: string}
+ *                         : (
+ *                             $limit is 3
+ *                                 ? array{0: string, 1?: string, 2?: string}
+ *                                 : non-empty-list<string>
```

Comment:
Note: I stopped at `3` array keys, but this could be made bigger and uglier in future.

For now, I think this is a decent trade-off to help most developers (most `explode()` calls provide a limit of `2`, and you should check for array key existence  regardless)

Reply:
`list{0: string, 1?: string, 2?: string}` would be better here (it would carry the meaning that the `2` offset can't exists if `1` does not

Reply:
Oooooh, I didn't know about this syntax, thanks!

Reply:
c0c01168091ef8a2eadf8b06532f303b480d6686